dev/system-test/firestore.ts

```diff
@@ -2447,6 +2458,89 @@ describe('WriteBatch class', () => {
+    await ref.set({number: FieldValue.increment(1)}, {merge: true});
+    await ref.set({number: FieldValue.delete()}, {merge: true});
+    await ref.set({address: {zip: FieldValue.delete()}}, {merge: true});
+    await ref.set(
+      {address: {nestedStringArray: FieldValue.delete()}},
+      {merge: true}
+    );
+    await ref.set(
+      {address: {nestedStringArray: FieldValue.arrayUnion('baz')}},
+      {merge: true}
+    );
+    await ref.set(
+      {address: {nestedStringArray: FieldValue.arrayRemove('baz')}},
+      {merge: true}
+    );
+
+    // TODO: Allow nested FieldValue
+    await ref.update({address: {nestedStringArray: FieldValue.delete()}});
+    await ref.update({
+      address: {nestedStringArray: FieldValue.arrayUnion('baz')},
+    });
+
+    // These should error.
+    await ref.set({numberArray: ['not-number']}, {merge: true});
+    await ref.set({number: 'not-number'}, {merge: true});
+    await ref.set({number: null}, {merge: true});
+    await ref.set({address: {nestedStringArray: 'not-array'}}, {merge: true});
+    await ref.set(
+      {address: {nestedStringArray: FieldValue.serverTimestamp()}},
+      {merge: true}
+    );
+
+    // NumericFieldValue Allowed
+    await ref.set({number: FieldValue.increment(1)}, {merge: true});
+    await ref.set({number: FieldValue.delete()}, {merge: true});
+
+    // NumericFieldValue Error
+    await ref.set({number: FieldValue.arrayUnion(4)}, {merge: true});
+    await ref.set({number: FieldValue.arrayRemove(4)}, {merge: true});
+    await ref.set({number: FieldValue.serverTimestamp()}, {merge: true});
+
+    // TimestampFieldValue Allowed
+    await ref.set({timestamp: FieldValue.serverTimestamp()}, {merge: true});
+    await ref.set({timestamp: FieldValue.delete()}, {merge: true});
+
+    // TimestampFieldValue Error
+    await ref.set({timestamp: FieldValue.increment(1)}, {merge: true});
+    await ref.set({timestamp: FieldValue.arrayUnion(4)}, {merge: true});
+    await ref.set({timestamp: FieldValue.arrayRemove(4)}, {merge: true});
+    await ref.set({timestamp: 3}, {merge: true});
+
+    // ArrayFieldValue Allowed
+    await ref.set({numberArray: FieldValue.arrayUnion(123)}, {merge: true});
+    await ref.set({numberArray: FieldValue.arrayRemove(12345)}, {merge: true});
+    await ref.set({numberArray: FieldValue.delete()}, {merge: true});
+
+    // ArrayFieldValue Error
+    await ref.set({numberArray: FieldValue.increment(12345)}, {merge: true});
+    await ref.set({numberArray: FieldValue.serverTimestamp()}, {merge: true});
+
+    // Non-special type Allowed
+    await ref.set({name: FieldValue.delete()}, {merge: true});
+
+    // Non-special type Error
+    await ref.set({name: FieldValue.increment(12345)}, {merge: true});
+    await ref.set({name: FieldValue.serverTimestamp()}, {merge: true});
+    await ref.set({name: FieldValue.arrayUnion(123)}, {merge: true});
+    await ref.set({name: FieldValue.arrayRemove(12345)}, {merge: true});
+  });
```

Comment:
You should add some tests that fail validation with your new code - https://dev.to/oliverjumpertz/making-good-use-of-ts-expect-error-in-typescript-1f41